[Bug] delete sql will fail when creating a table without explicitly declaring the number of buckets.

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.


### Paimon version

0.7

### Compute Engine

flink-1.17.2

### Minimal reproduce step

```
CREATE TABLE  if not EXISTS use_be_hours_4 (
    user_id BIGINT,
    item_id BIGINT,
    behavior STRING,
    dt STRING,
    hh STRING,
    PRIMARY KEY (user_id) NOT ENFORCED
) 
with (
  'manifest.format'='orc',
  'changelog-producer' = 'lookup',
  'incremental-between-scan-mode'='changelog',
  'changelog-producer.row-deduplicate'='true'
  
);


INSERT INTO use_be_hours_4
VALUES
    (1, 1, 'watch', '2022-01-01', '10'),
    (2, 2, 'like', '2022-01-01', '10'),
     (0, 0, 'watch', '2022-01-01', '10'); 

-- ERROR HAPPEN AS BELOW     
delete from use_be_hours_4 where user_id =2;
```

ERROR MSG:
```
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Can't extract bucket from row in dynamic bucket mode, you should use 'TableWrite.write(InternalRow row, int bucket)' method.
	at org.apache.paimon.table.TableUtils.deleteWhere(TableUtils.java:74)
	at org.apache.paimon.flink.sink.SupportsRowLevelOperationFlinkTableSink.executeDeletion(SupportsRowLevelOperationFlinkTableSink.java:175)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:889)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:874)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:991)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:765)
	at org.dinky.executor.DefaultTableEnvironment.executeSql(DefaultTableEnvironment.java:300)
	at org.dinky.executor.Executor.executeSql(Executor.java:208)
	at org.dinky.job.builder.JobDDLBuilder.run(JobDDLBuilder.java:47)
	at org.dinky.job.JobManager.executeSql(JobManager.java:339)
	... 136 more
Caused by: java.lang.IllegalArgumentException: Can't extract bucket from row in dynamic bucket mode, you should use 'TableWrite.write(InternalRow row, int bucket)' method.
	at org.apache.paimon.table.sink.DynamicBucketRowKeyExtractor.bucket(DynamicBucketRowKeyExtractor.java:44)
	at org.apache.paimon.table.sink.TableWriteImpl.toSinkRecord(TableWriteImpl.java:148)
	at org.apache.paimon.table.sink.TableWriteImpl.writeAndReturn(TableWriteImpl.java:125)
	at org.apache.paimon.table.sink.TableWriteImpl.write(TableWriteImpl.java:116)
	at org.apache.paimon.table.TableUtils.deleteWhere(TableUtils.java:67)
	... 145 more
```

However, if I explicitly declare the number of buckets when creating the table as in the statement below, the issue does not occur.

```
CREATE TABLE  if not EXISTS use_be_hours_4 (
    user_id BIGINT,
    item_id BIGINT,
    behavior STRING,
    dt STRING,
    hh STRING,
    PRIMARY KEY (user_id) NOT ENFORCED
) 
with (
  'manifest.format'='orc',
  'changelog-producer' = 'lookup',
  'incremental-between-scan-mode'='changelog',
  'changelog-producer.row-deduplicate'='true',
  'bucket' = '1'
  
);
```


### What doesn't meet your expectations?

I can run successfully regardless of whether I explicitly declare the number of buckets when creating the table (as it defaults to one bucket if not declared).



### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/docs/layouts/shortcodes/generated/core_configuration.html
+++ b/docs/layouts/shortcodes/generated/core_configuration.html
@@ -36,7 +36,7 @@ under the License.
             <td><h5>bucket</h5></td>
             <td style="word-wrap: break-word;">-1</td>
             <td>Integer</td>
-            <td>Bucket number for file store.</td>
+            <td>Bucket number for file store.<br />It should either be equal to -1 (dynamic bucket mode), or it must be greater than 0 (fixed bucket mode).</td>
         </tr>
         <tr>
             <td><h5>bucket-key</h5></td>

--- a/paimon-common/src/main/java/org/apache/paimon/CoreOptions.java
+++ b/paimon-common/src/main/java/org/apache/paimon/CoreOptions.java
@@ -74,7 +74,13 @@ public class CoreOptions implements Serializable {
             key("bucket")
                     .intType()
                     .defaultValue(-1)
-                    .withDescription("Bucket number for file store.");
+                    .withDescription(
+                            Description.builder()
+                                    .text("Bucket number for file store.")
+                                    .linebreak()
+                                    .text(
+                                            "It should either be equal to -1 (dynamic bucket mode), or it must be greater than 0 (fixed bucket mode).")
+                                    .build());
 
     @Immutable
     public static final ConfigOption<String> BUCKET_KEY =

--- a/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java
+++ b/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java
@@ -207,9 +207,9 @@ public interface FileIO extends Serializable {
 
     /** Read file to UTF_8 decoding. */
     default String readFileUtf8(Path path) throws IOException {
-        try (SeekableInputStream in = newInputStream(path)) {
-            BufferedReader reader =
-                    new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
+        try (SeekableInputStream in = newInputStream(path);
+                BufferedReader reader =
+                        new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
             StringBuilder builder = new StringBuilder();
             String line;
             while ((line = reader.readLine()) != null) {
@@ -233,8 +233,9 @@ public interface FileIO extends Serializable {
         Path tmp = path.createTempPath();
         boolean success = false;
         try {
-            try (PositionOutputStream out = newOutputStream(tmp, false)) {
-                OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+            try (PositionOutputStream out = newOutputStream(tmp, false);
+                    OutputStreamWriter writer =
+                            new OutputStreamWriter(out, StandardCharsets.UTF_8)) {
                 writer.write(content);
                 writer.flush();
             }
@@ -254,8 +255,8 @@ public interface FileIO extends Serializable {
      * implementations.
      */
     default void overwriteFileUtf8(Path path, String content) throws IOException {
-        try (PositionOutputStream out = newOutputStream(path, true)) {
-            OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+        try (PositionOutputStream out = newOutputStream(path, true);
+                OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8)) {
             writer.write(content);
             writer.flush();
         }
@@ -337,9 +338,9 @@ public interface FileIO extends Serializable {
         List<IOException> ioExceptionList = new ArrayList<>();
 
         // load preferIO
-        FileIOLoader perferIOLoader = config.preferIO();
+        FileIOLoader preferIOLoader = config.preferIO();
         try {
-            loader = checkAccess(perferIOLoader, path, config);
+            loader = checkAccess(preferIOLoader, path, config);
         } catch (IOException ioException) {
             ioExceptionList.add(ioException);
         }
@@ -403,10 +404,10 @@ public interface FileIO extends Serializable {
         if (loader == null) {
             String fallbackMsg = "";
             String preferMsg = "";
-            if (perferIOLoader != null) {
+            if (preferIOLoader != null) {
                 preferMsg =
                         " "
-                                + perferIOLoader.getClass().getSimpleName()
+                                + preferIOLoader.getClass().getSimpleName()
                                 + " also cannot access this path.";
             }
             if (fallbackIO != null) {

--- a/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java
+++ b/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java
@@ -207,9 +207,9 @@ public interface FileIO extends Serializable {
 
     /** Read file to UTF_8 decoding. */
     default String readFileUtf8(Path path) throws IOException {
-        try (SeekableInputStream in = newInputStream(path);
-                BufferedReader reader =
-                        new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
+        try (SeekableInputStream in = newInputStream(path)) {
+            BufferedReader reader =
+                    new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
             StringBuilder builder = new StringBuilder();
             String line;
             while ((line = reader.readLine()) != null) {
@@ -233,9 +233,8 @@ public interface FileIO extends Serializable {
         Path tmp = path.createTempPath();
         boolean success = false;
         try {
-            try (PositionOutputStream out = newOutputStream(tmp, false);
-                    OutputStreamWriter writer =
-                            new OutputStreamWriter(out, StandardCharsets.UTF_8)) {
+            try (PositionOutputStream out = newOutputStream(tmp, false)) {
+                OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
                 writer.write(content);
                 writer.flush();
             }
@@ -255,8 +254,8 @@ public interface FileIO extends Serializable {
      * implementations.
      */
     default void overwriteFileUtf8(Path path, String content) throws IOException {
-        try (PositionOutputStream out = newOutputStream(path, true);
-                OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8)) {
+        try (PositionOutputStream out = newOutputStream(path, true)) {
+            OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
             writer.write(content);
             writer.flush();
         }

--- a/paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java
+++ b/paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java
@@ -87,6 +87,8 @@ public class SchemaValidation {
 
         CoreOptions options = new CoreOptions(schema.options());
 
+        validateBucket(schema, options);
+
         validateDefaultValues(schema);
 
         validateStartupMode(options);
@@ -156,18 +158,6 @@ public class SchemaValidation {
                                             f, KEY_FIELD_PREFIX));
                         });
 
-        if (options.bucket() == -1 && options.toMap().get(BUCKET_KEY.key()) != null) {
-            throw new RuntimeException(
-                    "Cannot define 'bucket-key' in unaware or dynamic bucket mode.");
-        }
-
-        if (options.bucket() == -1
-                && schema.primaryKeys().isEmpty()
-                && options.toMap().get(FULL_COMPACTION_DELTA_COMMITS.key()) != null) {
-            throw new RuntimeException(
-                    "AppendOnlyTable of unware or dynamic bucket does not support 'full-compaction.delta-commits'");
-        }
-
         if (schema.primaryKeys().isEmpty() && options.streamingReadOverwrite()) {
             throw new RuntimeException(
                     "Doesn't support streaming read the changes from overwrite when the primary keys are not defined.");
@@ -180,17 +170,7 @@ public class SchemaValidation {
             }
         }
 
-        if (schema.crossPartitionUpdate()) {
-            if (options.bucket() != -1) {
-                throw new IllegalArgumentException(
-                        String.format(
-                                "You should use dynamic bucket (bucket = -1) mode in cross partition update case "
-                                        + "(Primary key constraint %s not include all partition fields %s).",
-                                schema.primaryKeys(), schema.partitionKeys()));
-            }
-        }
-
-        if (options.mergeEngine() == CoreOptions.MergeEngine.FIRST_ROW) {
+        if (options.mergeEngine() == MergeEngine.FIRST_ROW) {
             if (options.changelogProducer() != ChangelogProducer.LOOKUP) {
                 throw new IllegalArgumentException(
                         "Only support 'lookup' changelog-producer on FIRST_MERGE merge engine");
@@ -522,7 +502,7 @@ public class SchemaValidation {
                                 field);
                     });
 
-            if (options.mergeEngine() == CoreOptions.MergeEngine.FIRST_ROW) {
+            if (options.mergeEngine() == MergeEngine.FIRST_ROW) {
                 throw new IllegalArgumentException(
                         "Do not support use sequence field on FIRST_MERGE merge engine.");
             }
@@ -532,6 +512,32 @@ public class SchemaValidation {
                         String.format(
                                 "You can not use sequence.field in cross partition update case "
                                         + "(Primary key constraint '%s' not include all partition fields '%s').",
+                                schema.primaryKeys(), schema.partitionKeys()));
+            }
+        }
+    }
+
+    private static void validateBucket(TableSchema schema, CoreOptions options) {
+        int bucket = options.bucket();
+        if (bucket == -1) {
+            if (options.toMap().get(BUCKET_KEY.key()) != null) {
+                throw new RuntimeException(
+                        "Cannot define 'bucket-key' in unaware or dynamic bucket mode.");
+            }
+
+            if (schema.primaryKeys().isEmpty()
+                    && options.toMap().get(FULL_COMPACTION_DELTA_COMMITS.key()) != null) {
+                throw new RuntimeException(
+                        "AppendOnlyTable of unware or dynamic bucket does not support 'full-compaction.delta-commits'");
+            }
+        } else if (bucket < 1) {
+            throw new RuntimeException("The number of buckets needs to be greater than 0.");
+        } else {
+            if (schema.crossPartitionUpdate()) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "You should use dynamic bucket (bucket = -1) mode in cross partition update case "
+                                        + "(Primary key constraint %s not include all partition fields %s).",
                                 schema.primaryKeys(), schema.partitionKeys()));
             }
         }

--- a/paimon-core/src/test/java/org/apache/paimon/schema/TableSchemaTest.java
+++ b/paimon-core/src/test/java/org/apache/paimon/schema/TableSchemaTest.java
@@ -179,6 +179,25 @@ public class TableSchemaTest {
                 .hasMessageContaining("Field fake_col can not be found in table schema.");
     }
 
+    @Test
+    public void testBucket() {
+        List<DataField> fields =
+                Arrays.asList(
+                        new DataField(0, "f0", DataTypes.INT()),
+                        new DataField(1, "f1", DataTypes.INT()),
+                        new DataField(2, "f2", DataTypes.INT()));
+        List<String> partitionKeys = Collections.singletonList("f0");
+        List<String> primaryKeys = Collections.singletonList("f1");
+        Map<String, String> options = new HashMap<>();
+
+        TableSchema schema =
+                new TableSchema(1, fields, 10, partitionKeys, primaryKeys, options, "");
+
+        options.put(BUCKET.key(), "-2");
+        assertThatThrownBy(() -> validateTableSchema(schema))
+                .hasMessageContaining("The number of buckets needs to be greater than 0.");
+    }
+
     static RowType newRowType(boolean isNullable, int fieldId) {
         return new RowType(
                 isNullable,

--- a/paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java
+++ b/paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java
@@ -20,6 +20,7 @@ package org.apache.paimon.flink.sink;
 
 import org.apache.paimon.flink.log.LogStoreTableFactory;
 import org.apache.paimon.table.Table;
+import org.apache.paimon.table.sink.BatchTableCommit;
 
 import org.apache.flink.table.catalog.ObjectIdentifier;
 import org.apache.flink.table.connector.sink.abilities.SupportsTruncate;
@@ -41,6 +42,10 @@ public class FlinkTableSink extends SupportsRowLevelOperationFlinkTableSink
 
     @Override
     public void executeTruncation() {
-        table.newBatchWriteBuilder().newCommit().truncateTable();
+        try (BatchTableCommit batchTableCommit = table.newBatchWriteBuilder().newCommit()) {
+            batchTableCommit.truncateTable();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
     }
 }

--- a/paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/SupportsRowLevelOperationFlinkTableSink.java
+++ b/paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/SupportsRowLevelOperationFlinkTableSink.java
@@ -58,6 +58,7 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.stream.Collectors;
 
+import static org.apache.paimon.CoreOptions.BUCKET;
 import static org.apache.paimon.CoreOptions.MERGE_ENGINE;
 
 /** Flink table sink that supports row level update and delete. */
@@ -193,7 +194,8 @@ public abstract class SupportsRowLevelOperationFlinkTableSink extends FlinkTable
     }
 
     private boolean canPushDownDeleteFilter() {
-        return deletePredicate == null || deleteIsDropPartition() || deleteInSingleNode();
+        return -1 != Options.fromMap(table.options()).get(BUCKET)
+                && (deletePredicate == null || deleteIsDropPartition() || deleteInSingleNode());
     }
 
     private boolean deleteIsDropPartition() {

--- a/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/DynamicBucketTableITCase.java
+++ b/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/DynamicBucketTableITCase.java
@@ -57,7 +57,7 @@ public class DynamicBucketTableITCase extends CatalogITCaseBase {
     }
 
     @Test
-    public void testWriteRead() {
+    public void testWriteReadDelete() {
         sql("INSERT INTO T VALUES (1, 1, 1), (1, 2, 2), (1, 3, 3), (1, 4, 4), (1, 5, 5)");
         assertThat(sql("SELECT * FROM T"))
                 .containsExactlyInAnyOrder(
@@ -74,6 +74,10 @@ public class DynamicBucketTableITCase extends CatalogITCaseBase {
                         Row.of(1, 3, 33),
                         Row.of(1, 4, 4),
                         Row.of(1, 5, 5));
+        sql("DELETE FROM T WHERE pt = 1 AND pk = 3");
+        assertThat(sql("SELECT * FROM T"))
+                .containsExactlyInAnyOrder(
+                        Row.of(1, 1, 11), Row.of(1, 2, 2), Row.of(1, 4, 4), Row.of(1, 5, 5));
 
         assertThat(sql("SELECT DISTINCT bucket FROM T$files"))
                 .containsExactlyInAnyOrder(Row.of(0), Row.of(1));

